graphs/graph_modules/data_fetcher.py

```diff
@@ -0,0 +1,365 @@
+
+"""
+Data fetching and caching functionality for TGraph Bot.
+Handles both synchronous and asynchronous data retrieval from Tautulli API.
+"""
+
+from cachetools import TTLCache
+from datetime import datetime, timedelta
+from typing import Dict, Any, Optional
+import aiohttp
+import asyncio
+import logging
+import requests
+
+class DataFetcherError(Exception):
+    """Base exception for data fetching errors."""
+    pass
+
+class ApiError(DataFetcherError):
+    """Raised when there's an error in API communication."""
+    pass
+
+class DataProcessingError(DataFetcherError):
+    """Raised when there's an error processing the fetched data."""
+    pass
+
+class DataFetcher:
+    """Handles fetching and caching of data from Tautulli API."""
+    
+    def __init__(self, config: Dict[str, Any], cache_ttl: int = 300):
+        """
+        Initialize the DataFetcher.
+        
+        Args:
+            config: Configuration dictionary containing API settings and parameters
+            cache_ttl: Cache time-to-live in seconds (default: 300)
+        """
+        self.config = config
+        self._cache_ttl = cache_ttl
+        self._api_cache = TTLCache(maxsize=100, ttl=cache_ttl)
+
+    def _create_cache_key(self, cmd: str, params: Dict[str, Any]) -> tuple:
+        """
+        Create a consistent cache key from command and parameters.
+        
+        Args:
+            cmd: API command
+            params: Command parameters
+            
+        Returns:
+            A tuple suitable for use as a cache key
+        """
+        sorted_params = tuple(sorted(
+            (k, str(v)) for k, v in params.items() 
+            if v is not None
+        ))
+        return (cmd, sorted_params)
+
+    def _get_cached_data(self, cache_key: tuple) -> Optional[Dict[str, Any]]:
+        """
+        Retrieve data from cache.
+        
+        Args:
+            cache_key: Cache key tuple
+            
+        Returns:
+            Cached data if available, None otherwise
+        """
+        return self._api_cache.get(cache_key)
+
+    def _set_cached_data(self, cache_key: tuple, data: Dict[str, Any]) -> None:
+        """
+        Store data in cache.
+        
+        Args:
+            cache_key: Cache key tuple
+            data: Data to cache
+        """
+        self._api_cache[cache_key] = data
+
+    def _validate_data_structure(self, cmd: str, data: Any) -> bool:
+        """
+        Validate the data structure returned from Tautulli based on the command.
+        
+        Args:
+            cmd: The API command
+            data: The data to validate
+            
+        Returns:
+            True if the data structure is valid, False otherwise
+        """
+        if cmd == "get_users":
+            return True  # get_users has a different structure, handled separately
+            
+        if not isinstance(data, dict):
+            return True  # Raw API responses are processed differently for some commands
+
```

Comment:
The current implementation is actually correct and well-designed. Returning True for non-dict data types is not a bug but a feature that allows for proper handling of varied API response formats. Changing this as suggested in the review would break functionality for legitimate API responses that don't return dictionary structures.